### PR TITLE
IMP: add ruff and codespell  to pre-commit hooks

### PR DESCRIPTION
**Pull request by antoinecollet5**
See issue #145 
IMP: add ruff, codespell and isort to pre-commit hooks + do required corrections to comply with the tools

Note: I haven't add mypy since it is already covered by tox. But it could be an option in a second PR.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,3 +1,16 @@
 [flake8]
 ignore = E302, W503, E501, E741, E203, F405
 max-line-length = 88
+
+[codespell]
+skip = *.pyc,*.gif,*.png,*.jpg,*.ply, ./bibliography.bib,*.ipynb
+ignore-words-list = lod,byteorder,flem,parm,doubleclick,revered,inout,fro,nd,sies,hist,ans
+quiet-level = 3
+
+[tool.isort]
```
Comment (dafeda) [inline setup.cfg:10]: Do we need `isort` when using `ruff`? Based on my understanding, `ruff` replaces `isort`.

Comment (antoinecollet5) [reply setup.cfg]: @dafeda indeed I was not aware that ruff now includes isort. I made the changes and squashed my commits.

**Review by tommyod — APPROVED** [review]
Two small concerns:

- Maintaining codespell's `ignore-words-list` might become a pain. But I'm very willing to test out codespell. If it ends up being a pain in the future, we could always remove codespell from the pre commit and just run it manually from time to time.
- The cost of having two formatters instead of one. Ruff seems like it does relatively little extra compared to black, but again - let's try it out. It has 18k stars compared to blacks 34k, so it's very popular.

Overall I'm positive to merge and try this out :)

**Comment by dafeda** [comment]
> Two small concerns:
> 
> * Maintaining codespell's `ignore-words-list` might become a pain. But I'm very willing to test out codespell. If it ends up being a pain in the future, we could always remove codespell from the pre commit and just run it manually from time to time.
> * The cost of having two formatters instead of one. Ruff seems like it does relatively little extra compared to black, but again - let's try it out. It has 18k stars compared to blacks 34k, so it's very popular.
> 
> Overall I'm positive to merge and try this out :)

`ruff` is not meant to replace `black` but instead `flake8`, `isort` and a bunch of others.
Here are the "Rules" `ruff` supports: https://docs.astral.sh/ruff/rules/

**Comment by dafeda** [comment]
@antoinecollet5 I suggest we get rid of flake8 now that we have ruff. Do you agree?

**Comment by antoinecollet5** [comment]
@dafeda I agree that flake8 is no longer needed. Do you mean removing the config from setup.cfg ?

And indeed, ruff is not overlapping with black yet. I think there are some experimental versions including black. So somewhere in the future we might remove black as well.

**Comment by tommyod** [comment]
> @dafeda I agree that flake8 is no longer needed. Do you mean removing the config from setup.cfg ?

I think that's what he means. Searching [flake8](https://github.com/search?q=repo%3Aequinor%2Fiterative_ensemble_smoother%20flake8&type=code) in the repo gives no other results.

**Comment by dafeda** [comment]
> > @dafeda I agree that flake8 is no longer needed. Do you mean removing the config from setup.cfg ?
> 
> I think that's what he means. Searching [flake8](https://github.com/search?q=repo%3Aequinor%2Fiterative_ensemble_smoother%20flake8&type=code) in the repo gives no other results.

Yes 👍

**Comment by antoinecollet5** [comment]
@dafeda @tommyod  Done for flake8

**Review by dafeda — APPROVED** [review]

**Comment by dafeda** [comment]
Thanks @antoinecollet5 , good work 👍

**Pull request merged**